### DISCUSSION TITLE
RFC:Support user-defined functions

### DISCUSSION THREAD

**Discussion by zhaojunnana** [vesoft-inc/nebula#4793] [General]
In complex queries, the udf function is sometimes necessary.
Here are some of my ideas and simple implementations: https://github.com/zhaojunnana/nebula/blob/master/docs/UDF.md
Because we use the Java technology stack, we consider using JNI at the beginning of implementation. Graphd calls the Java jar. The reason for giving up is that the performance loss is considerable.

pr:https://github.com/vesoft-inc/nebula/pull/4795

**Comment by Shylock-Hg** [comment]
lgtm for me, but I think there should be an abstraction about UDF implementation (environment or handler?). So user could choose execute UDF by specified runtime (JNI in your RFC).

**Reply by Shylock-Hg** [reply]
In other words, should support to configuration UDF runtime(JNI in your RFC or others in futrhe).

**Comment by Shylock-Hg** [comment]
https://github.com/vesoft-inc/nebula/issues/4060
A previous request maybe could help you.

**Comment by xtcyclist** [comment]
So you've tried to use java for udf? But gave it up later because of performance losses?

**Comment by zhaojunnana** [comment]
是的，关于设计和当时性能测试
<img width="1035" alt="image" src="https://user-images.githubusercontent.com/44078666/198204591-2968d524-1e7c-4a71-82b7-d943434cc239.png">

**Reply by xtcyclist** [reply]
👍 测得很清楚。性能确实是差得比较多。

**Reply by codesigner** [reply]
在java/c++上交互上性能的确不好做, 目前的内存layout可能做不到指针传递,  堆外内存等方式, 改动会很大
